### PR TITLE
Surround with try/catch reformats document

### LINKED ISSUE #1572: Surround with try/catch reformats document
This behaviour is more then a year old, but it often bugs me so hard that I thought it's time for a report.
It just happened again on a really large file where I noticed it way too late.

Ok, when using try/catch the whole document gets reformatted:

before:
Screen Shot 2020-08-11 at 22 00 03

after:
Screen Shot 2020-08-11 at 22 00 13

##### Environment
- Operating System: OSX
- JDK version: OpenJDK 11
- Visual Studio Code version: latest
- Java extension version: latest

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #1572 
Requires https://github.com/eclipse/eclipse.jdt.ls/pull/1551

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -543,6 +543,18 @@
...
+          "description": "Insert spaces when pressing Tab. Default is inherited from `editor.insertSpaces`",
+          "default": null,
+          "scope": "window"
+        },
+        "java.editor.tabSize": {
+          "type": "integer",
+          "description": "Tab Size. Default is inherited from `editor.tabSize`",
+          "default": null,
+          "scope": "window"
+        },
```
Comment (fbricon) [inline package.json:557]: I don't think it makes sense to expose new  tab settings to users, it'll be more confusing than anything. Why would anyone want to use different tab settings than the ones from vscode?

File: README.md [hunk 2]
```diff
@@ -173,11 +173,13 @@ The following settings are supported:
...
 * `java.project.resourceFilters`: Excludes files and folders from being refreshed by the Java Language Server, which can improve the overall performance. For example, ["node_modules",".git"] will exclude all files and folders named 'node_modules' or '.git'. Defaults to ["node_modules",".git"].
-
-New in 0.67.0:
 * `java.templates.fileHeader`: Specifies the file header comment for new Java file. Supports configuring multi-line comments with an array of strings, and using ${variable} to reference the [predefined variables](https://github.com/redhat-developer/vscode-java/wiki/Predefined-Variables-for-Java-Template-Snippets).
 * `java.templates.typeComment`: Specifies the type comment for new Java type. Supports configuring multi-line comments with an array of strings, and using ${variable} to reference the [predefined variables](https://github.com/redhat-developer/vscode-java/wiki/Predefined-Variables-for-Java-Template-Snippets).
 
+New in 0.68.0:
+* `java.editor.insertSpaces`: Insert spaces when pressing Tab. Default is inherited from `editor.insertSpaces`.
+* `java.editor.tabSize`: Tab Size. Default is inherited from `editor.tabSize`.
+
```
Comment (fbricon) [inline README.md:182]: No need to expose those settings

File: src/extension.ts [hunk 3]
```diff
@@ -394,6 +394,13 @@ export function getJavaConfig(javaHome: string) {
 	const origConfig = getJavaConfiguration();
 	const javaConfig = JSON.parse(JSON.stringify(origConfig));
 	javaConfig.home = javaHome;
+	const editorConfig = workspace.getConfiguration('editor');
+	if (javaConfig.editor.insertSpaces === null) {
+		javaConfig.editor.insertSpaces = editorConfig.get('insertSpaces');
```
Comment (fbricon) [inline src/extension.ts:399]: javaConfig.format.insertSpaces

File: src/extension.ts [hunk 4]
```diff
@@ -394,6 +394,13 @@ export function getJavaConfig(javaHome: string) {
...
+	}
+	if (javaConfig.editor.tabSize === null) {
+		javaConfig.editor.tabSize = editorConfig.get('tabSize');
```
Comment (fbricon) [inline src/extension.ts:402]: javaConfig.format.tabSize

**Review by fbricon — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (snjeza) [reply README.md]: Fixed

File: package.json (hunk 1 above)
Comment (snjeza) [reply package.json]: Fixed.

File: src/extension.ts (hunk 4 above)
Comment (snjeza) [reply src/extension.ts]: Fixed

File: src/extension.ts (hunk 3 above)
Comment (snjeza) [reply src/extension.ts]: Fixed

**Comment by fbricon** [comment]
unfortunately, if vscode's default setting is set to use spaces/4, but the document contains only tab/4, then adding new lines manually will have vscode automatically detect the indentation and insert tabs, but applying a code action will insert spaces.
That's the one good thing the previous code had, i.e. maintain consistent editing experience.

I think we'll need to apply https://github.com/redhat-developer/vscode-java/pull/1081/ to make it useable.

**Comment by snjeza** [comment]
> I think we'll need to apply #1081 to make it useable.

Agree.

**Comment by testforstephen** [comment]
There is another PR https://github.com/eclipse/eclipse.jdt.ls/pull/1657 to use #1081, so no need to take care of it in this PR.

@snjeza could you rebase and resolve the conflicts?

**Comment by snjeza** [comment]
> @snjeza could you rebase and resolve the conflicts?

Done.

File: src/extension.ts [hunk 5]
```diff
@@ -403,6 +403,9 @@ export function getJavaConfig(javaHome: string) {
 	// Since output path is a project specific setting. To avoid pollute other project,
 	// we avoid reading the value from the global scope.
 	javaConfig.project.outputPath = origConfig.inspect<string>("project.outputPath").workspaceValue;
+	const editorConfig = workspace.getConfiguration('editor');
+	javaConfig.format.insertSpaces = editorConfig.get('insertSpaces');
+	javaConfig.format.tabSize = editorConfig.get('tabSize');
```
Comment (testforstephen) [inline src/extension.ts:408]: You should add "editor.insertSpaces" and "editor.tabSize" to [the synchronized configuration sections](https://github.com/redhat-developer/vscode-java/blob/master/src/extension.ts#L163) as well.

image

Comment (snjeza) [reply src/extension.ts:408]: Fixed.

**Review by testforstephen — APPROVED** [review]

**Pull request merged**